android/app/src/main/java/com/zulipmobile/sharing/ReceiveShareActivity.kt

```diff
@@ -0,0 +1,129 @@
+package com.zulipmobile.sharing
+
+import android.content.Context
+import android.content.Intent
+import android.content.res.Configuration
+import android.net.Uri
+import android.os.Bundle
+import android.os.Parcelable
+import android.util.Log
+import android.webkit.WebView
+import androidx.annotation.Nullable
+import com.facebook.react.ReactActivity
+import com.facebook.react.ReactApplication
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.bridge.WritableMap
+import com.facebook.react.bridge.Arguments
+import com.zulipmobile.MainActivity
+import com.zulipmobile.notifications.*
+
+const val TAG = "ZulipReceiveShare"
+
+class ReceiveShareActivity : ReactActivity() {
+
+    /**
+     * Returns the name of the main component registered from JavaScript.
+     * This is used to schedule rendering of the component.
+     */
+    override fun getMainComponentName(): String? {
+        return "SharingRoot"
+    }
+
+    private fun sendEvent(reactContext: ReactContext,
+                          eventName: String,
+                          @Nullable params: WritableMap) {
+        Log.d(TAG, "Sending event with shared data")
+        emit(reactContext, eventName, params)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        WebView.setWebContentsDebuggingEnabled(true)
+        val application = application as ReactApplication
+        launchMainActivity(application as Context)
```

Comment:
Comparing this logic with `NotifyReact.kt`, one difference I notice is that this one unconditionally calls `launchMainActivity`, while that one only calls it if the app isn't already in the foreground.

Is that difference intentional? What's the effect of it?

Reply:
The logic is still different from what's in `NotifyReact.kt`, but in a different way:
```kt
        if (null == reactContext) {
            launchMainActivity(application as Context)
            SharingModule.initialSharedData = params
        } else {
            sendEvent(reactContext, "shareReceived", params)
        }
```

If you compare to the logic in `notifyReact`, you'll find that that existing logic will call `launchMainActivity` if the status is NOT_RUNNING or BACKGROUND, and this won't. It doesn't sound like that's intentional.

This logic is pretty tricky and interacts with subtle details of Android and of RN. For that kind of logic, it's best to make a point of not writing it twice.

Ultimately the way to do that is to refactor the existing code so it can be used in both situations. Or, for development to first get the thing working, a fine shortcut can be to copy the code and make only the needed changes (like the event name "notificationOpened"), with the intent to go back and factor it out after we understand exactly what we need.